src/devices/obs.ts

```diff
@@ -0,0 +1,740 @@
+
+		// store the new state, for later use:
+		this.setState(newOBSState, newState.time)
+	}
+
+	clearFuture (clearAfterTime: number) {
+		// Clear any scheduled commands after this time
+		this._doOnTime.clearQueueAfter(clearAfterTime)
+	}
+
+	async terminate () {
+		this._setDisconnected = true
+		this._doOnTime.dispose()
+		this._obs.disconnect()
+		return Promise.resolve(true)
+	}
+
+	getStatus (): DeviceStatus {
+		let statusCode = StatusCode.GOOD
+		let messages: Array<string> = []
+
+		if (!this._connected) {
+			statusCode = StatusCode.BAD
+			messages.push('Not connected')
+		} else if (!this._authenticated) {
+			statusCode = StatusCode.BAD
+			messages.push('Not authenticated')
+		}
+
+		return {
+			statusCode: statusCode,
+			messages: messages,
+			active: this.isActive
+		}
+	}
+
+	async makeReady (okToDestroyStuff?: boolean): Promise<void> {
+		if (okToDestroyStuff) {
+			// do something?
+		}
+	}
+
+	get canConnect (): boolean {
+		return !this._connected && !this._retryConnectTimeout && !this._setDisconnected
+	}
+
+	get connected (): boolean {
+		return this._connected
+	}
+
+	convertStateToOBS (state: TimelineState, mappings: Mappings): OBSState {
+		if (!this._initialized) {
+			throw Error(
+				'convertStateToOBS cannot be used before inititialized'
+			)
+		}
+
+		let deviceState = this._getDefaultState()
+
+		// Sort layer based on Mapping type (to make sure audio is after inputs) and Layer name
+		const sortedLayers = _.sortBy(
+			_.map(state.layers, (tlObject, layerName) => {
+				const tlObjectExt = tlObject as ResolvedTimelineObjectInstanceExtended
+				let mapping: MappingOBS = mappings[layerName] as MappingOBS
+				if (!mapping && tlObjectExt.isLookahead && tlObjectExt.lookaheadForLayer) {
+					mapping = mappings[tlObjectExt.lookaheadForLayer] as MappingOBS
+				}
+				return {
+					layerName,
+					tlObject,
+					mapping
+				}
+			}).sort((a, b) => a.layerName.localeCompare(b.layerName)),
+			(o) => o.mapping.mappingType
+		)
+
+		_.each(sortedLayers, ({ tlObject, mapping }) => {
+			if (mapping) {
+				switch (mapping.mappingType) {
+					case MappingOBSType.CurrentScene:
+						if (
+							tlObject.content.type ===
+							TimelineContentTypeOBS.CURRENT_SCENE
+						) {
+							if ((tlObject as ResolvedTimelineObjectInstanceExtended).isLookahead) {
+								let obsTlCurrentScene = (tlObject as any) as TimelineObjOBSCurrentScene
+								deviceState.previewScene =
+									obsTlCurrentScene.content.sceneName
+							} else {
+								let obsTlCurrentScene = (tlObject as any) as TimelineObjOBSCurrentScene
+								deviceState.currentScene =
+									obsTlCurrentScene.content.sceneName
+							}
+						}
+						break
+					case MappingOBSType.CurrentTransition:
+						if (
+							tlObject.content.type ===
+							TimelineContentTypeOBS.CURRENT_TRANSITION
+						) {
+							let obsTlCurrentTransition = (tlObject as any) as TimelineObjOBSCurrentTransition
```

Comment:
what if this was a lookahead object via `lookaheadForLayer`? Those objects are essentially background layer objects, which here will change the 'foreground' `currentTransition` property.

Same concern with the rest of the mappings handling below